talisker/logs.py

```diff
@@ -341,54 +349,100 @@ def logfmt_atoms(self, structured):
-                if key is not None:
+                if key is None:
+                    key_errors.append(k)
+                else:
                     yield key, '"???"'
 
+        if key_errors:
+            # we embed the keys in the message precisely because they
+            # couldn't be parsed as logfmt key
+            logger = logging.getLogger(__name__)
+            logger.warning('could not parse logfmt keys: ' + str(key_errors))
+
     def logfmt_key(self, k):
         if isinstance(k, bytes):
             k = k.decode('utf8')
 
-        # TODO: look at optimizing this to avoid string allocations
+        if isinstance(k, str):
+            k = k.replace(' ', '_').replace('.', '_').replace('=', '_')
+            k = self.safe_string(k, self.MAX_KEY_SIZE, self.TRUNCATED_KEY)
+            # TODO: look at measuring perf of this
+            # ' ' and = are replaced because they're are not valid logfmt
+            # . is replaced because elasticsearch can't do keys with . in
+        elif isinstance(k, bool):
+            # need to do this here, as bool are also numbers
+            return None
+        elif isinstance(k, numbers.Number):
+            k = str(k)
+        else:
+            return None
 
-        k = k.strip()[:self.MAX_KEY_SIZE]
-        # ' ' and = are replaced because they're are not valid logfmt
-        # . is replaced because elasticsearch can't do keys with . in
-        k = k.replace(' ', '_').replace('.', '_').replace('=', '_')
-        # strip " as grok parser can not escape them
-        k = k.replace('"', '')
         return k
 
     def logfmt_value(self, v):
-        if isinstance(v, bytes):
-            v = v.decode('utf8')
+        """Format a python value for logfmt.
 
-        if isinstance(v, str):
-            v = v.strip()
-            # logstash kv filter cannot handle escaped "
-            v = v.replace('"', '')
+        The output target here is influenced by logstash/elastic search types.
 
-            if len(v) > self.MAX_VALUE_SIZE:
-                v = v[:self.MAX_VALUE_SIZE] + self.TRUNCATED
+        For strings, ensures unicode, quotes, and truncates as needed.
+        For bools, format as json truth values.
+        For dicts, truncate.
+        For others types (e.g. numbers) leave as is.
+        For all else, coerce to string.
+        """
 
+        if isinstance(v, bytes):
+            v = v.decode('utf8')
+        if isinstance(v, str):
+            v = self.safe_string(v, self.MAX_VALUE_SIZE, self.TRUNCATED)
+            # explicitly quote strings to avoid ambiguity
             v = '"' + v + '"'
         elif isinstance(v, bool):
             v = str(v).lower()
-        elif isinstance(v, dict):
-            return '"' + self.TRUNCATED + '"'
+        elif isinstance(v, numbers.Number):
+            v = str(v)
+        else:
+            v = '"' + str(type(v)) + '"'
 
         return v
 
+    def safe_string(self, s, max, truncate_str=None):
```

Comment:
why is `truncate_str` optional? you're passing it through for both the key and value cases..

Reply:
Yes, and artifact for a previous iteration, will remove